src/host/outputStream.cpp

```diff
@@ -299,6 +299,31 @@ BOOL ConhostInternalGetSet::PrivateSetExtendedTextAttributes(const ExtendedAttri
     return TRUE;
 }
 
+// Method Description:
+// - Retrieves the current TextAttribute of the active screen buffer.
+// Arguments:
+// - pAttrs: Receives the TextAttribute value.
+// Return Value:
+// - TRUE if successful. FALSE otherwise.
+BOOL ConhostInternalGetSet::PrivateGetTextAttributes(TextAttribute* const pAttrs) const

```

Comment:
I suppose this change was inevitable. We've kinda just been patching around not just exposing the entire TextAttribute through this layer, but this might be easier. I know there was previously reluctance to doing this, something about it exposing too much of the implementation details of how conhost itself was implemented. @miniksa might remember more. I think we've come a _long_ way since then, and this might be reasonable nowadays.

I guess from a design perspective, it would change things pretty dramatically - because the adapter could directly control the state of the text attributes, it could just avoid using any of the other functions, and put all the combined attribute manipulation logic directly in the adapter. There _might_ be other things that conhost needs to do during some of these calls that might get circumnavigated by just using this method. These things might be a little trickier to move up into the adapter. I haven't dug in on this too much, so it might not be a problem.

If it was possible to move this logic into the adapter, then we could probably get rid of a _bunch_ of these more specific boilerplate-y functions. That would probably make life easier for everyone.

The alternative here would be exposing a `PrivateSaveState`/`PrivateRestoreState` pair that lets conhost implement it however it likes. Though, when I think about it, the implementation would be _highly_ similar between conhost and Terminal...

Reply:
>If it was possible to move this logic into the adapter, then we could probably get rid of a bunch of these more specific boilerplate-y functions. That would probably make life easier for everyone.

Absolutely! I had exactly the same thought when I added this. There's a lot of weirdness in the current SGR implementation which I think could be greatly simplified if we had access to an API like this. Was going to raise a separate issue for that once I knew you were happy with the API and I had some time to play around with potential implementations.

> The alternative here would be exposing a PrivateSaveState/PrivateRestoreState pair that lets conhost implement it however it likes. Though, when I think about it, the implementation would be highly similar between conhost and Terminal...

If there was a need for conhost and Terminal to handle this functionality differently, that's already not a problem, since Terminal shares none of the `AdaptDispatch` implementation. That said, I'm not sure that's the right approach in the long term, because it seems we're slowly reimplementing all of the `AdaptDispatch` code in `TerminalDispatch` when those implementations should really be identical 99% of the time. But that's a topic for another issue.

Reply:
> > If it was possible to move this logic into the adapter, then we could probably get rid of a bunch of these more specific boilerplate-y functions. That would probably make life easier for everyone.
> 
> Absolutely! I had exactly the same thought when I added this. There's a lot of weirdness in the current SGR implementation which I think could be greatly simplified if we had access to an API like this. Was going to raise a separate issue for that once I knew you were happy with the API and I had some time to play around with potential implementations.

I'm OK with additional private methods being added here, I just don't want to expose any more publicly. @zadjii-msft, I'm not sure I had an extreme reticence to adding new private APIs given we've done a ton of them. I think it was more "I'm going to try to accomplish as much as possible without adding new APIs at all" and it led us to the non-ideal situations that we had and discovered over time.

> > The alternative here would be exposing a PrivateSaveState/PrivateRestoreState pair that lets conhost implement it however it likes. Though, when I think about it, the implementation would be highly similar between conhost and Terminal...
> 
> If there was a need for conhost and Terminal to handle this functionality differently, that's already not a problem, since Terminal shares none of the `AdaptDispatch` implementation. That said, I'm not sure that's the right approach in the long term, because it seems we're slowly reimplementing all of the `AdaptDispatch` code in `TerminalDispatch` when those implementations should really be identical 99% of the time. But that's a topic for another issue.

Yes, if we're duplicating a big portion of the dispatch... then we probably need to somehow abstract the buffers or whatever the two adapters are calling and make that a shared component. Feel free to file another issue for that if you feel you can articulate it.